### PR TITLE
introduce DagsterModel

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

Introduces a `StrictModel` class, which is a subclass of `pydantic.BaseModel` that's frozen and doesn't allow default constructor arguments that aren't class members.

The idea is for this to eventually replace all our uses of `NamedTuple`, `BaseModel`, and `@dataclass`.

This is a replacement for the unmerged https://github.com/dagster-io/dagster/pull/20461.

Downstream PRs that demonstrate its usage:
- https://github.com/dagster-io/dagster/pull/20641
- https://github.com/dagster-io/dagster/pull/20643
- https://github.com/dagster-io/dagster/pull/20638

### CODE REVIEW

**Comment by sryza** [comment]
Closing this while I work out some issues with inheritance

**Comment by sryza** [comment]
Ok so I have encountered a problem. This fails type checking:

```python
    class MyClass(StrictModel):
        foo: str
        bar: int

        def __init__(self, foo: str, bar: int):
            super().__init__(foo=foo, bar=bar)

    MyClass(foo="fdsjk", bar=4)
```

with
```
  10:29: No parameter named "foo" (reportCallIssue)

  10:38: No parameter named "bar" (reportCallIssue)
```

**Comment by schrockn** [comment]
Yeah I think you have to do:

```python
class StrictModel(BaseModel):
    def __init__(self, **data: Any) -> None:
        super().__init__(**data)

    class Config:
        extra = "forbid"
        frozen = True
```

**Comment by sryza** [comment]
@schrockn then pyright won't raise static type errors if if incorrect args are passed to the superclass constructor :(

E.g. this would ideally fail at type check time instead of runtime:
```python
    class MyClass(StrictModel):
        bar: int

        def __init__(self, bar: str):
            super().__init__(bar=bar)

    with pytest.raises(ValidationError):
        MyClass(bar="fdslk")
```

This doesn't leak to users though, so not the end of the world. Pushing up a version with this change.

**Comment by alangenfeld** [comment]
`Model` ? 
```from dagster._core.model import Model```  

We don't have anything that conflicts. Since this is internal to the codebase, I think unqualified `Model` communicates this is the *the* way to do things without the `Dagster` qualifier.  I could imagine having `MutableModel` as a sibling to highlight classes that go against immutable default. 

Considered  stuff like `Record` `DataObject` `Entity` but I think having `Model` in the name has the nice property of maintaining a connection to pydantic Model

`DagsterModel` is fine too

File: python_modules/dagster/dagster/_core/utils.py
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/utils.py:1]: nit: having this in `utils` gives it an optional vibe to me, would consider moving it to `models` / `types` or something

**Comment by schrockn** [comment]
My only concern with super generic names like "Model" is that auto import functionality in IDEs often provide you a whole litany of potential option. Other than that I support it. Like you I think we should stick with Model nomenclature to retain connection with pydantic.

**Comment by alangenfeld** [comment]
> auto import functionality in IDEs often provide you a whole litany of potential option

Good call out.

I did a quick test in my vscode set-up and didn't get any other hits for capital M Model 
Screenshot 2024-03-22 at 9 38 23 AM

Not sure how good of a proxy for other peoples experience that is though

**Comment by sryza** [comment]
> Like you I think we should stick with Model nomenclature to retain connection with pydantic.

+1

I see the case for both `DagsterModel` and `Model` and could be swayed either way. @schrockn @alangenfeld - are either of you swayed by the points the other one brought up?

**Comment by schrockn** [comment]
I’m good either way! Slight preference for “DagsterModel” since I think it kind of accurately states “this is how we have decided to use pydantic in Dagster” rather than implying that it’s a generic thing. However super lightly held and I’m perfectly happy with “Model”

File: python_modules/dagster/dagster/_core/utils.py [hunk 1]
```diff
@@ -196,3 +197,12 @@ def submit(self, fn, *args, **kwargs):
 def is_valid_email(email: str) -> bool:
     regex = r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,7}\b"
     return bool(re.fullmatch(regex, email))
+
+
+class StrictModel(BaseModel):
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/utils.py:202]: leave a lil doc block with some explanation of why standardizing on these settings

**Review by alangenfeld — APPROVED** [review]
`DagsterModel` it is

**Comment by sryza** [comment]
@schrockn @alangenfeld - pushed up a new version with `DagsterModel` under `_models` instead of `_core.utils`

**Pull request merged**